File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def deleteNode(self, root, key):
        """
        :type root: TreeNode
        :type key: int
        :rtype: TreeNode
        """
        """
            We connect the left branch to the parent of the key node and then we find the greatest node of that left branch and connect the right branch of key node to that node's right
        """
        def findLastRight(node): # Find the last rightmost node of the given node
            if not node.right:
                return node
            return findLastRight(node.right)
        
        def helper(node):
            if not node.left:
                return node.right
            if not node.right:
                return node.left
            right = node.right
            left = findLastRight(node.left)
            left.right = node.right
            return node.left
        
        
        
        if not root:
            return None
        
        if root.val == key:
            return helper(root)
        
        dummy = root
        
        while root:
            if root.val > key:
                if root.left and root.left.val == key:
                    root.left = helper(root.left)
                else:
                    root = root.left
            else:
                if root.right and root.right.val == key:
                    root.right = helper(root.right)
                else:
                    root = root.right
        
        return dummy
    
    

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return null;
        if (root.val == key) return helper(root);
        
        TreeNode cur = root;
        
        while (cur != null){
            if (cur.val > key){
                if ((cur.left != null) && (cur.left.val == key)){
                    cur.left = helper(cur.left);
                }
                else cur = cur.left;
            }
            else{
                if ((cur.right != null) && (cur.right.val == key)){
                    cur.right = helper(cur.right);
                }
                else cur = cur.right;
            }
        }
        
        return root;
    }
    
    private TreeNode helper(TreeNode node){
        if (node.left == null) return node.right;
        if (node.right == null) return node.left;
        
        TreeNode rightMost = findRightMost(node.left);
        rightMost.right = node.right;
        return node.left;
    }
    
    private TreeNode findRightMost(TreeNode node){
        if (node.right == null) return node;
        return findRightMost(node.right);
    }
}